**[RQ4] What is the most common way of payments? Discover the way payments are executed in each borough and visualize the number of payments for any possible means. Then run the Chi-squared test to see whether the method of payment is correlated to the borough. Then, comment the results.**

In [1]:
###month of MARCH

#We imported only the columns useful to answer our Rquestion.
#NB:Payment_type represented by a numeric code signifying how the passenger paid for the trip.  
#1= Credit card; 2= Cash; 3= No charge; 4= Dispute; 5= Unknown; 6= Voided trip.

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import math

data=pd.read_csv('/Users/Enzopc/Desktop/yellow_tripdata_2018-03.csv', usecols=['passenger_count','PULocationID', 'DOLocationID','payment_type','total_amount'])
zone=pd.read_csv('/Users/Enzopc/Desktop/taxi _zone_lookup.csv', usecols=['LocationID', 'Borough'])
datazone=pd.merge(data, zone, how='left', left_on=['PULocationID'],right_on=['LocationID'])

print("Number of rows:", datazone.shape[0])
print("Number of columns: ", datazone.shape[1])
datazone.head()

Number of rows: 9430376
Number of columns:  7


passenger_count  PULocationID  DOLocationID  payment_type  total_amount  \
0                1           145           145             2          3.80   
1                1           151           244             1         17.15   
2                1           238           152             2         11.30   
3                1           145           145             2          3.80   
4                1           145           145             2          4.30   

   LocationID    Borough  
0         145     Queens  
1         151  Manhattan  
2         238  Manhattan  
3         145     Queens  
4         145     Queens

In [2]:
# What is the most common way of payments? #
datazone['payment_type'].value_counts()

1    6560898
2    2804905
3      50978
4      13595
Name: payment_type, dtype: int64

As we can see from an initial count, it seems that the most common way of payments is by credit card


1: 6560898   2: 2804905    3: 50978    4: 13595

In [3]:
#But now let's see more deeply: 
#Discover the way payments are executed in each borough and visualize the number of payments for any possible means.

contingency_table = pd.crosstab(datazone['payment_type'], datazone['Borough'])
contingency_table

Borough       Bronx  Brooklyn  EWR  Manhattan  Queens  Staten Island  Unknown
payment_type                                                                 
1              4669     79629  493    5977982  386393             73   111659
2              5329     41518  114    2538900  170450             51    48543
3               267      1480   42      40555    6486             12     2136
4                65       310   10      11231    1524              3      452

**Chi-squared test to see whether the method of payment is correlated to the borough.**

H0 (Null Hypothesis): There is no relationship between variable one and variable two.

H1 (Alternative Hypothesis): There is a relationship between variable 1 and variable 2.

If the p-value is significant, you can reject the null hypothesis and claim that the findings support the alternative hypothesis.


Chi-square Assumptions that need to be met in order for the results of the Chi-square test to be trusted:

 -When testing the data, the cells should be frequencies or counts of cases and not percentages. It is okay to convert to percentages after testing the data
 
 -The levels (categories) of the variables being tested are mutually exclusive
 
 -Each participant contributes to only one cell within the Chi-square table
 
 -The groups being tested must be independent
 
 -The value of expected cells should be greater than 5

If all of these assumptions are met, then Chi-square is the correct test to use.

In [4]:
from scipy import stats

stats.chi2_contingency(contingency_table)

(13980.172855859311,
 0.0,
 18,
 array([[7.18678410e+03, 8.55296880e+04, 4.58479257e+02, 5.96139080e+06,
         3.92979338e+05, 9.67050330e+01, 1.13256204e+05],
        [3.07248286e+03, 3.65655204e+04, 1.96008345e+02, 2.54860461e+06,
         1.68005921e+05, 4.13431866e+01, 4.84191176e+04],
        [5.58411181e+01, 6.64563363e+02, 3.56237143e+00, 4.63198453e+04,
         3.05343883e+03, 7.51395490e-01, 8.79997640e+02],
        [1.48919142e+01, 1.77228195e+02, 9.50026277e-01, 1.23527462e+04,
         8.14302265e+02, 2.00384905e-01, 2.34680998e+02]]))

The first value (13980.17) is the Chi-square value, followed by the p-value ~ 0, then comes the degrees of freedom (18), and lastly it outputs the expected frequencies as an array. 
Since all of the expected frequencies are greater than 5, the Chi-square test results can be trusted.
We can reject the null hypothesis as the p-value is less than 0.05. 
Thus, the results indicate that there is a correlation/some sort of relationship between payment type and borough. For sure we do know that these two variables are not independent of each other.

We could relate this result to the "industrialization" of boroughs. Focusing on Manhattan, Brooklyn, EWR and Queens we can see that the most common way of payment is with credit card. Fos Staten Island there's no big difference and in Bronx most of the people pay with cash.

In [ ]:
###Grafico per visualizzare il numero di pagamenti per ogni mezzo###

#contingency_table = pd.crosstab(datazone['payment_type'], datazone['Borough'])
#Assigns the frequency values
#creditcard = contingency_table.iloc[0][0:6].values
#cash = contingency_table.iloc[1][0:6].values
#Plots the bar chart
#fig = plt.figure(figsize=(10, 5))
#sns.set(font_scale=1.5)
#categories = ["Bronx","Brooklyn","Manhattan","Queens","Staten Island","Unknown"]
#p1 = plt.bar(categories, creditcard, color='#d62728')
#p2 = plt.bar(categories, cash, bottom=creditcard)
#plt.legend((p2[0], p1[0]), ('Cash', 'Credit Card'))
#plt.xlabel('Borough')
#plt.ylabel('Count')
#plt.show()